pkg/scheduler/plugins/reservation/plugin.go

```diff
@@ -219,412 +190,165 @@ func (p *Plugin) Filter(ctx context.Context, cycleState *framework.CycleState, p
+func (pl *Plugin) Reserve(ctx context.Context, cycleState *framework.CycleState, pod *corev1.Pod, nodeName string) *framework.Status {
 	if reservationutil.IsReservePod(pod) {
 		return nil
 	}
 
-	// if the pod match a reservation
-	state := getPreFilterState(cycleState)
-	if state == nil || state.skip { // expected matchedCache is prepared
-		klog.V(5).InfoS("skip the Reservation Reserve", "pod", klog.KObj(pod), "node", nodeName)
-		return nil
-	}
-	rOnNode := state.matchedCache.GetOnNode(nodeName)
-	if len(rOnNode) <= 0 { // the pod is suggested to bind on a node with no reservation
+	nominatedReservation := frameworkext.GetNominatedReservation(cycleState)
+	if nominatedReservation == nil {
+		klog.V(5).Infof("Skip reserve with reservation since there are no matched reservations, pod %v, node: %v", klog.KObj(pod), nodeName)
 		return nil
 	}
 
-	// select one reservation for the pod to allocate
-	// sort: here we use MostAllocated (simply set all weights as 1.0)
-	var order int64 = math.MaxInt64
-	for i := range rOnNode {
-		var rInfo *reservationInfo
-		if order == math.MaxInt64 {
-			rInfo, order = findMostPreferredReservationByOrder(rOnNode)
-			if order > 0 && rInfo != nil {
-				rInfo.Score = mostPreferredScore
-				continue
-			}
-		}
-		rOnNode[i].ScoreForPod(pod)
-	}
-	sort.Slice(rOnNode, func(i, j int) bool {
-		return rOnNode[i].Score >= rOnNode[j].Score
-	})
-
-	// NOTE: matchedCache may be stale, try next reservation when current one does not match any more
-	// TBD: currently Reserve got a failure if any reservation is selected but all failed to reserve
-	for _, rInfo := range rOnNode {
-		target := rInfo.GetReservation()
-		// use the cached reservation, in case the version in cycle state is too old/incorrect or mutated by other pods
-		rInfo = p.reservationCache.GetInCache(target)
-		if rInfo == nil {
-			// check if the reservation is marked as expired
-			if p.reservationCache.IsInactive(target) { // in case reservation is marked as inactive
-				klog.V(5).InfoS("skip reserve current reservation since it is marked as expired",
-					"pod", klog.KObj(pod), "reservation", klog.KObj(target))
-			} else {
-				klog.V(4).InfoS("failed to reserve current reservation since it is not found in cache",
-					"pod", klog.KObj(pod), "reservation", klog.KObj(target))
-			}
-			continue
-		}
-
-		// avoid concurrency conflict inside the scheduler (i.e. scheduling cycle vs. binding cycle)
-		if !matchReservation(pod, rInfo) {
-			klog.V(5).InfoS("failed to reserve reservation since the reservation does not match the pod",
-				"pod", klog.KObj(pod), "reservation", klog.KObj(target), "reason", dumpMatchReservationReason(pod, rInfo))
-			continue
-		}
+	// NOTE: Having entered the Reserve stage means that the Pod scheduling is successful,
+	// even though the associated Reservation may have expired, but in fact the real impact
+	// will not be encountered until the next round of scheduling.
+	assumed := nominatedReservation.DeepCopy()
```

Comment:
When multiple reservations are consumed in a pod, only one reservation is recorded, which may cause inconsistencies between the scheduler accounting and reservation status in some cases.

Reply:
Most of the scenes use AllocateOnce when setting up a Reservation, which is basically consistent with the Resource Requests of Pods and generally there will be no fragmentation. When AllocateOnce=false, a new constraint is added: only one Reservation with the same Owner can be scheduled on the same node, and when returning the resources reserved by the Reservation (with AllocateOnce=false), only the one with the earliest creation time is selected.